src/domain/core/src/repos/dataset_repository.rs

```diff
@@ -433,6 +461,8 @@ pub enum GetDatasetError {
     #[error(transparent)]
     NotFound(#[from] DatasetNotFoundError),
     #[error(transparent)]
+    MultiTenantRefUnexpected(#[from] MultiTenantRefUnexpectedError),
```

Comment:
Looking further in the PR I feel even strongly about this - `MultiTenantRefUnexpected` ended up being passed through in almost every other error type, but I really don't think it deserves the same "fundamental error class" status as `Unauthorized` or `Internal`.

Reply:
I agree that it's a usage mistake and it would make sense to report it in methods like `resolve_dataset_ref()`, to let caller handle it.

But when some high-level call like `sync_dataset()` crashes with `MultiTenantRefUnexpected` - it might as well crash with `InternalError` because caller will have no means to handle this error - such error should've been prevented earlier at input validation.

So overall I think:
- It's a scary amount of changes in error handling for this small error
- I would be OK to see internal error on tenancy issues in CLI as they shouldn't happen during normal usage
- If we really want to display `MultiTenantRefUnexpected` as user error - we could use RTTI to detect it even when it's wrapped into `InternalError`
- But the best solution would probably be to prevent it from happening by validating user input early (but I'm not sure extra core is worth it in this specific case)

Reply:
> But the best solution would probably be to prevent it from happening by validating user input early (but I'm not sure extra core is worth it in this specific case)

That's a right comment. I've migrated the `MultiTenantRefUnexpectedError` error to CLI error and implemented validation at the level of command lines, where we expect `DatasetRef`, not `DatasetRefAny`.